sys/kern/kern_cheri_revoke.c

```diff
@@ -489,6 +489,14 @@ kern_cheri_revoke(struct thread *td, int flags, cheri_revoke_epoch start_epoch,
 			 * so that we don't race any page faults from kernel
 			 * worker threads; we won't race any page faults from
 			 * userspace already since we're single-threaded.
+			 *
+			 * XXXMJ this statement is not quite true.  The map lock
+			 * is acquired during page fault handing but may be
+			 * dropped (and re-acquired) around calls into the
+			 * pager.  In other words, it appears to be possible to
+			 * increment the GCLG after a (kernel) thread has
+			 * faulted but before it has installed a PTE for the new
+			 * mapping.
```

Comment:
That's a fair point.  Is it sufficient that those faults hold the `vm_map_lock` across the call to `pmap_enter`?  If so, does holding the pmap lock in `pmap_caploadgen_next` suffice and the `vm_map_lock` here is just unnecessary and/or confusion on my part?

Reply:
I confess that I do not fully understand the synchronization requirements here. Is the purpose of the vm_map lock here only to serialize loads of the pmap's GCLG? I note that pmap_enter() loads the GCLG before acquiring the pmap lock, so it seems that the pmap lock on its own is not sufficient.